core/src/main/clojure/xtdb/buffer_pool.clj

```diff
@@ -35,258 +48,424 @@
+  (write [_ src] (.write file-channel src)))
+
+(defrecord DiskOnlyBufferPool
+  [allocator
+   ^ArrowBufLRU memory-store
+   ^Path disk-store]
+  Closeable
+  (close [_]
+    (free-memory memory-store)
+    (util/close allocator))
   IBufferPool
   (getBuffer [_ k]
-    (CompletableFuture/completedFuture
-     (when k
-       (if-let [cached-buffer (cache-get buffers k)]
-         (do
-           (record-cache-hit cached-buffer)
-           cached-buffer)
-
-         (let [nio-buffer (os/get-path root-path k)
-               create-arrow-buf #(util/->arrow-buf-view allocator nio-buffer)
-               [_ buf] (cache-compute buffers k create-arrow-buf)]
-           buf)))))
-
-  (getRangeBuffer [_ k start len]
-    (object-store/ensure-shared-range-oob-behaviour start len)
-
-    (CompletableFuture/completedFuture
-     (when k
-       (if-let [cached-buffer (or (cache-get buffers [k start len])
-                                  (when-let [^ArrowBuf cached-full-buffer (cache-get buffers k)]
-                                    (.slice cached-full-buffer start len)))]
-         (do
-           (record-cache-hit cached-buffer)
-           cached-buffer)
-
-         (let [nio-buffer (os/get-path-range root-path k start len)
-               create-arrow-buf #(util/->arrow-buf-view allocator nio-buffer)
-               [_ buf] (cache-compute buffers [k start len] create-arrow-buf)]
-           buf)))))
-
-  (putObject [_ k buf] (CompletableFuture/completedFuture (os/put-path root-path k buf)))
-  (listObjects [_this] (os/list-path root-path))
-  (listObjects [_this dir] (os/list-path root-path dir))
+    (let [cached-buffer (cache-get memory-store k)]
+      (cond
+        (nil? k)
+        (CompletableFuture/completedFuture nil)
+
+        cached-buffer
+        (do (record-cache-hit cached-buffer)
+            (CompletableFuture/completedFuture cached-buffer))
 
+        :else
+        (let [buffer-cache-path (.resolve disk-store (str k))]
+          (-> (if (util/path-exists buffer-cache-path)
+                ;; todo could this not race with eviction? e.g exists for this cond, but is evicted before we can map the file into the cache?
+                (CompletableFuture/completedFuture buffer-cache-path)
+                (CompletableFuture/failedFuture (os/obj-missing-exception k)))
+              (util/then-apply
+                (fn [path]
+                  (let [nio-buffer (util/->mmap-path path)
+                        create-arrow-buf #(util/->arrow-buf-view allocator nio-buffer)
+                        [_ buf] (cache-compute memory-store k create-arrow-buf)]
+                    buf))))))))
+  (listObjects [_]
+    (with-open [dir-stream (Files/walk disk-store (make-array FileVisitOption 0))]
+      (vec (sort (for [^Path path (iterator-seq (.iterator dir-stream))
+                       :when (Files/isRegularFile path (make-array LinkOption 0))]
+                   (str (.relativize disk-store path)))))))
+  (listObjects [_ dir]
```

Comment:
can we reinstate the blank lines removed between different methods here? without the whitespace I find my eyes don't have anything to lock on to when scanning the code